### PR TITLE
ghdl: add support for llvm backend

### PR DESCRIPTION
**Pull request by timor**
###### Motivation for this change

This adds support for building ghdl with the llvm backend, which enables among other things, to
have a working version for 64 bit systems.  Unfortunately it only works in 16.03, not master, since
gnat is broken there.
###### Things done

Added new attribute "ghdl_llvm" in addition to "ghdl_mcode", by introducing the argument "flavour"
to the ghdl nix expression and choosing the build flags according to the specified flavour.
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Make the existing ghdl recipe more flexible, and introduce "ghdl_llvm"
as a package in addition to "ghdl_mcode".  This seems to specifically
require llvm 3.5, though.

cc @viric

### CODE REVIEW

File: pkgs/development/compilers/ghdl/default.nix [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
-# I think that mcode can only generate x86 code,
-# so it fails to link pieces on x86_64.
-assert stdenv.system == "i686-linux";
+# mcode only works on x86, while the llvm flavour works on both x86 and x86_64.
+
+
+assert flavour == "llvm" || flavour == "mcode";
 
 let
+  optional = stdenv.lib.optional;
```
Comment (joachifm) [inline pkgs/development/compilers/ghdl/default.nix]: Equivalently, `inherit (stdenv.lib) optional`

File: pkgs/development/compilers/ghdl/default.nix [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
   version = "0.33";
 in
 stdenv.mkDerivation rec {
-  name = "ghdl-mcode-${version}";
+  name = "ghdl-${version}";
```
Comment (joachifm) [inline pkgs/development/compilers/ghdl/default.nix]: In keeping with the original name, maybe embed the flavour in the name?

File: pkgs/development/compilers/ghdl/default.nix [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 
   src = fetchurl {
```
Comment (joachifm) [inline pkgs/development/compilers/ghdl/default.nix]: While you're at it, maybe port this to `fetchFromGitHub`, it should be more stable when using github's generated tarballs.

File: pkgs/development/compilers/ghdl/default.nix [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
     url = "https://github.com/tgingold/ghdl/archive/v${version}.tar.gz";
     sha256 = "09yvgqyglbakd74v2dgr470clzm744i232nixyffcds55vkij5da";
   };
 
-  buildInputs = [ gnat zlib ];
+  buildInputs = [ gnat zlib ] ++ optional (flavour == "llvm") [ clang ncurses ];
```
Comment (joachifm) [inline pkgs/development/compilers/ghdl/default.nix]: I think you want `optionals` here; `optional cond x == [ x ]` if `cond` is true.

**Comment by vcunat** [comment]
Note: gnat got fixed for unstable/master and 16.09 (see 36ef583e47).

**Comment by timor** [comment]
Works against master now, updated pull request: #18566

**Pull request closed without merging**